### PR TITLE
wlr-data-device

### PR DESCRIPTION
**Pull request by acrisci**
In this PR, I'm going to make the clipboard actually work.

I'm going to go a different way with this interface to make drag and drop a bit easier to do later.

TODO:

* [x] basic clipboard
* [x] drag and drop
* [x] drag and drop icons
* [x] handle multiple data devices bound to a seat
* [ ] <s>xwayland clipboard sync</s> (needs another pr)
* [ ] <s>cleanup (refactor seat functions into seat?)</s> (needs another pr)
* [x] seat selection signal

### CODE REVIEW

File: include/wlr/interfaces/wlr_data_source.h [hunk 1]
```diff
@@ -1,16 +0,0 @@
-#ifndef WLR_INTERFACES_WLR_DATA_SOURCE_H
-#define WLR_INTERFACES_WLR_DATA_SOURCE_H
-
-#include <wlr/types/wlr_data_source.h>
-
-struct wlr_data_source_impl {
-	void (*send)(struct wlr_data_source *data_source, const char *type, int fd);
-	void (*accepted)(struct wlr_data_source *data_source, const char *type);
-	void (*cancelled)(struct wlr_data_source *data_source);
-};
```
Comment (nyorain) [inline include/wlr/interfaces/wlr_data_source.h]: This was there to allow non-wayland data sources on compositor side. Would be needed for x11 <-> wayland clipboard synchronization and supporting stuff like the sway set_clipboard command (or whatever a compositor wants to do with the clipboard).

Comment (acrisci) [reply include/wlr/interfaces/wlr_data_source.h]: Oh I see thanks.

**Comment by acrisci** [comment]
Basic clipboard is pretty much done and ready to be tested but this probably needs some cleanup.

**Comment by SirCmpwn** [comment]
Tested it and it seems to work.

**Comment by acrisci** [comment]
Basic drag and drop is ready to be tested. You can use the gtk test program `testdnd2` to test it.

**Comment by acrisci** [comment]
Ok I believe this is ready to go.

The clipboard sync (and related refactoring) is kind of a lot of work, so I'll leave it to another PR. I want to make it easy for compositors (as well as xwayland) to get/set selection contents so those goals will share a lot of code.

Test with `weston-dnd`. The gtk test is buggy and I don't think they are my bugs.

I am aware that the icon is not shown in the intended place, but that is an unrelated issue.

**Comment by SirCmpwn** [comment]
I tried this with weston-dnd and it did not work.

**Comment by SirCmpwn** [comment]
testdnd2 crashes as well

**Comment by acrisci** [comment]
Oops, 09f16058 broke it. That commit disables client cursors which I accidently committed earlier because client cursors screw everything up for me.

**Comment by acrisci** [comment]
Ok it works again. Ready to go.

**Comment by SirCmpwn** [comment]
Maybe I'm missing something. I ran weston-dnd and tried to drag things around. Then I ran nautilus and tried to drag things around.

**Comment by SirCmpwn** [comment]
Neither worked.

**Comment by SirCmpwn** [comment]
testdnd2 (gtk+) didn't work either.

**Comment by emersion** [comment]
Tested `weston-dnd`  and `gnome-calculator` (dragging text between two windows) with Wayland backend, works.

**Comment by SirCmpwn** [comment]
Okay, tested this again.

DRM backend:

- weston-dnd - does not work, hangs weston-dnd
- testdnd2 - does not work, hangs testdnd2
- nautilus - does not work, but does not hang

Wayland backend: All three work correctly

**Comment by acrisci** [comment]
I can reproduce this segfault on my laptop that uses the nouveau driver, so it is probably related to the video card driver. Confirmed working for amd and intel drivers.

**Comment by SirCmpwn** [comment]
Segfault?

**Comment by acrisci** [comment]
Oh nevermind I was on the wrong branch. It works perfectly on my laptop with nouveau so that's not the issue.

**Comment by acrisci** [comment]
I don't know what to do about this. I can't reproduce the issue. Can you find out where it hangs for you in weston-dnd? Sometimes it hangs when it is expecting a frame callback.

**Comment by SirCmpwn** [comment]
I will investigate this some more myself today. Right now I'm working on putting a bow on #264

**Comment by SirCmpwn** [comment]
Can you figure out the conflicts on this? I'm just going to merge it and sort out my problems separately, since no one can reproduce them.

**Comment by acrisci** [comment]
It should be good to go.

**Pull request merged**